components/server/src/user/user-controller.ts

```diff
@@ -281,6 +286,95 @@ export class UserController {
                 res.sendStatus(200);
             },
         );
+
+        router.post(
+            "/auth/sendHeartbeat/:instanceID",
+            async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+                if (!req.isAuthenticated() || !User.is(req.user)) {
+                    res.sendStatus(401);
+                    log.warn("unauthenticated send heartbeat", { instanceId: req.params.instanceID });
+                    return;
+                }
+
+                const user = req.user as User;
+                if (user.blocked) {
+                    res.sendStatus(403);
+                    log.warn("blocked user attempted to send heartbeat", {
+                        instanceId: req.params.instanceID,
+                        userId: user.id,
+                    });
+                    return;
+                }
+
+                const wasClosed = req.body.wasClosed;
+                // const roundTripTime = req.body.roundTripTime;
+                const instanceID = req.params.instanceID;
+                if (!instanceID) {
+                    res.sendStatus(400);
+                    log.warn("attempted to send heartbeat without instance ID", {
+                        instanceId: req.params.instanceID,
+                        userId: user.id,
+                    });
+                    return;
+                }
+                const [workspace, instance] = await Promise.all([
```

Comment:
is it copied from somewhere? can we extract it to some service and reuse in both places that code of sending heartbeating is not duplicated

Reply:
It's a copy from code above, we can refactor the code but not in this PR?

Reply:
I think we should refactor in this PR. copy/pasting is not good.

Reply:
Refactor should go with add middlewares like userVerify, workspaceOwnedVerify etc,

It'll increase the difficulty of PR testing, and maybe break something, that all these APIs are very important

Reply:
We need to verify these things for this endpoint as well, no?